Update MoveIt Setup Assistant documentation for ROS Humble (#647)
ros-planning/moveit2_tutorials: doc/examples/setup_assistant/setup_assistant_tutorial.rst

MoveIt Setup Assistant
========================

.. image:: setup_assistant_launch.png
   :width: 700px
   :align: center

Overview
--------
The MoveIt Setup Assistant is a graphical user interface for configuring any robot for use with MoveIt.
Its primary function is generating a Semantic Robot Description Format (SRDF) file for your robot,
which specifies additional information required by MoveIt such as planning groups, end-effectors, and various kinematic parameters.
Additionally, it generates other necessary configuration files for use with the MoveIt pipeline.
To use the MoveIt Setup Assistant, you will need to have a URDF file for your robot.

Once you have a URDF file, you can open the MoveIt Setup Assistant and import your URDF.
This tutorial will guide you through a series of steps to configure various aspects of your robot,
such as defining its kinematic structure, specifying planning groups and end-effectors, and collision checking related setting.
To learn more about the URDF and SRDF, you can refer to the `URDF SRDF Overview <https://moveit.picknik.ai/humble/doc/examples/urdf_srdf/urdf_srdf_tutorial.html#>`_ page.

Getting Started
---------------

MoveIt and ROS 2

> * Follow the `instructions for installing MoveIt <https://moveit.picknik.ai/humble/doc/tutorials/getting_started/getting_started.html>`_

Reviewer: For all these types of URLs, instead of using the absolute links, you can do relative links like this:

```
:doc:`instructions for installing MoveIt </doc/tutorials/getting_started/getting_started>`
```

Where the link is the relative path of the `.rst` file in the repo directory.

This makes it so that when you switch doc versions, the right link is maintained instead of always pointing to Humble.

Author: Okay, I understand the reason now. Thank you! 

The URLs use :doc: now to refer to other files in the documentation.

  first if you have not already done that.

* We use the `moveit_resources_panda_description <https://github.com/ros-planning/moveit_resources/tree/humble/panda_description/urdf>`_

>   package for Humble. This package should already be included in your workspace if you've completed the MoveIt installation instructions.

Reviewer: For external links, you can keep the URL as is. However, try not to mention a specific release. I'd just point this to master and not mention Humble here.

Author: The tutorial is updated not to mention a specific release, but I think we should keep the link pointing to the Humble branch because the master is already changed. 

I made it like that because in the Humble release documentation, Getting Started Tutorial, [Download Source Code of MoveIt and the Tutorials](https://moveit.picknik.ai/humble/doc/tutorials/getting_started/getting_started.html#download-source-code-of-moveit-and-the-tutorials) section mentions the branch explicitly as follows.

```
cd ~/ws_moveit2/src
git clone https://github.com/ros-planning/moveit2_tutorials -b humble --depth 1
``` 
If we point to the master here, things will be different for the users. For example, `panda.urdf` is changed to `panda_arm_hand.urdf.xacro`.

I am aware of this point and will maintain the tutorial to keep things consistent. If you have any suggestions, please share with me and I will be happy to consider them.

Step 1: Start
-------------

* To start the MoveIt Setup Assistant: ::

   ros2 launch moveit_setup_assistant setup_assistant.launch.py

* This will bring up the start screen with two choices: **Create New
  MoveIt Configuration Package** or **Edit Existing MoveIt
  Configuration Package**.

* Click on the **Create New MoveIt Configuration Package** button to
  bring up the following screen:

.. image:: setup_assistant_create_package.png
   :width: 700px
   :align: center

* Click on the **Browse** button and navigate to the *panda.urdf* file

>   from the *moveit_resources_panda_description package* available in the following path: :: 

Reviewer: For code, use double backticks to it's formatted as `panda.urdf` instead of _panda.urdf_

   :width: 700px
   :align: center

Add the arm group

* We will first add Panda arm as a planning group

  * Enter **Group Name** as ``panda_arm``.

  * Choose **kdl_kinematics_plugin/KDLKinematicsPlugin** as the
    kinematics solver. *Note: if you have a custom robot and would
    like a powerful custom IK solver, see* `Kinematics/IKFast <https://moveit.picknik.ai/humble/doc/examples/ikfast/ikfast_tutorial.html>`_

  * Let **Kin. Search Resolution** and **Kin. Search Timeout** stay at
    their default values.

.. image:: planning_groups/setup_assistant_panda_arm_group.png
   :width: 700px
   :align: center

* Now, click on the **Add Joints** button. You will see a
  list of joints on the left-hand side. You need to choose all the
  joints that belong to the arm and add them to the right-hand
  side. The joints are arranged in the order that they are stored in
  an internal tree structure. This makes it easy to select a serial
  chain of joints.

  * Click on ``virtual_joint``, hold down the **Shift**
    button on your keyboard and then click on the
    ``panda_joint8``. Now click on the **>** button to add these
    joints into the list of **Selected Joints** on the right.

> .. image:: planning_groups/setup_assistant_panda_arm_group_joints.png 

Reviewer: there is an extra whitespace at the end of this

Its primary function is generating a Semantic Robot Description Format (SRDF) file for your robot,
which specifies additional information required by MoveIt such as planning groups, end effectors, and various kinematic parameters.
Additionally, it generates other necessary configuration files for use with the MoveIt pipeline.
To use the MoveIt Setup Assistant, you will need to have a URDF file for your robot.

Once you have a URDF file, you can open the MoveIt Setup Assistant and import your URDF.
This tutorial will guide you through a series of steps to configure various aspects of your robot,
such as defining its kinematic structure, specifying planning groups and end effectors, and collision checking related settings.
To learn more about the URDF and SRDF, you can refer to the `URDF SRDF Overview <https://moveit.picknik.ai/humble/doc/examples/urdf_srdf/urdf_srdf_tutorial.html#>`_ page.

Getting Started
---------------

MoveIt and ROS 2

* Follow the :doc:`instructions for installing MoveIt </doc/tutorials/getting_started/getting_started>`
  first if you have not already done that.

> * We use the `moveit_resources_panda_description <https://github.com/ros-planning/moveit_resources/tree/humble/panda_description/urdf>`_

Author: I appreciate your suggestion! The link is shortened.

such as the arm or end effector, to facilitate motion planning.

A move group can be configured to correspond to a specific kinematic chain on the robot,
which is a set of links and joints that define a sequence of transformations
from the base of the robot to the end effector. For example, a move group might be defined to
represent the arm of a robot, which would consist of all the links and joints necessary to move the arm.

Move groups can also be represented by sets of links or joints corresponding on the robot.
For example, a move group might be defined to represent the gripper of a robot,
which would consist of all the links or joints necessary to move together to realize a gripper opening or closing motion.

* Click on the **Planning Groups** pane selector.

* Click on **Add Group** and you should see the following screen:

.. image:: planning_groups/setup_assistant_panda_planning_groups.png
   :width: 700px
   :align: center

Add the arm group

* We will first add Panda arm as a planning group

  * Enter **Group Name** as ``panda_arm``.

  * Choose **kdl_kinematics_plugin/KDLKinematicsPlugin** as the

>     kinematics solver. Note: if you have a custom robot and would

Reviewer 2: Consider adding an actual RST `.. note::` here.

For example, end effectors can be used for attaching objects to the arm while carrying out pick-and-place tasks.

* Click on the **End Effectors** pane.

* Click **Add End Effector**.

* Choose ``hand`` as the **End Effector Name** for the gripper.

* Select ``hand`` as the **End Effector Group**.

* Select ``panda_link8`` as the **Parent Link** for this end-effector.

* Leave **Parent Group** blank.

.. image:: setup_assistant_panda_add_end_effector.png
   :width: 700px

Step 7: Add Passive Joints
--------------------------

The **Passive Joints** pane is meant to allow specification of any passive
joints that might exist in a robot. These are joints that are unactuated,
meaning that they cannot be directly controlled. It's important to specify
passive joints so that the planners are aware of their existence and can avoid
planning for them. If the planners do not know about the passive joints, they
might try to plan trajectories that involve moving the passive joints, which would
result in invalid plans. The Panda robot arm does not have any passive joints so we will skip this step.

> Step 8: ros2_control URDF Modification

Reviewer 2: Suggest adding a RST `.. note::` here that lets the user know if their robots URDF/xacro already has a ros2_control.xacro included that this step can be skipped.

Author: Thank you for pointing that out. It will help prevent a lot of confusion.

MoveIt Setup Assistant
========================

.. image:: setup_assistant_launch.png
   :width: 700px
   :align: center

Overview
--------
The MoveIt Setup Assistant is a graphical user interface for configuring any robot for use with MoveIt.
Its primary function is generating a Semantic Robot Description Format (SRDF) file for your robot,

> which specifies additional information required by MoveIt such as planning groups, end-effectors, and various kinematic parameters.

Reviewer: ```suggestion
which specifies additional information required by MoveIt such as planning groups, end effectors, and various kinematic parameters.
```

Additionally, it generates other necessary configuration files for use with the MoveIt pipeline.
To use the MoveIt Setup Assistant, you will need to have a URDF file for your robot.

Once you have a URDF file, you can open the MoveIt Setup Assistant and import your URDF.
This tutorial will guide you through a series of steps to configure various aspects of your robot,

> such as defining its kinematic structure, specifying planning groups and end-effectors, and collision checking related setting.

Reviewer: ```suggestion
such as defining its kinematic structure, specifying planning groups and end effectors, and collision checking related settings.
```

> To learn more about the URDF and SRDF, you can refer to the `URDF SRDF Overview <https://moveit.picknik.ai/humble/doc/examples/urdf_srdf/urdf_srdf_tutorial.html#>`_ page.

Reviewer: Change to `:doc:` tags here.

Author: Sorry about missing this one!

Getting Started
---------------

MoveIt and ROS 2

* Follow the :doc:`instructions for installing MoveIt </doc/tutorials/getting_started/getting_started>`
  first if you have not already done that.

* We use the `moveit_resources_panda_description <https://github.com/ros-planning/moveit_resources/tree/humble/panda_description/urdf>`_
  package. This package should already be included in your workspace if you've completed the MoveIt installation instructions.

Step 1: Start
-------------

* To start the MoveIt Setup Assistant: ::

   ros2 launch moveit_setup_assistant setup_assistant.launch.py

* This will bring up the start screen with two choices: **Create New
  MoveIt Configuration Package** or **Edit Existing MoveIt
  Configuration Package**.

* Click on the **Create New MoveIt Configuration Package** button to
  bring up the following screen:

.. image:: setup_assistant_create_package.png
   :width: 700px
   :align: center

* Click on the **Browse** button and navigate to the ``panda.urdf`` file
  from the ``moveit_resources_panda_description package`` available in the following path: ::

   ~/ws_moveit2/src/moveit_resources/panda_description/urdf/panda.urdf

> 
> 

Reviewer: ```suggestion

```

   :width: 700px
   :align: center

Step 3: Add Virtual Joints
--------------------------
Virtual joints are primarily used to connect robots to the world.
In the case of the Panda arm, which is a fixed base manipulator, defining a fixed virtual joint
is optional. However, we will define a ``fixed`` virtual joint that attaches the ``panda_link0``
of the arm to the ``world`` frame. This virtual joint signifies that the base of the arm
remains stationary in the world frame.

* Click on the **Virtual Joints** pane selector. Click on **Add Virtual Joint**.

* Set the joint name as ``virtual_joint``.

* Set the child link as ``panda_link0`` and the parent frame name as ``world``.

* Set the Joint Type as ``fixed``.

* Click **Save** and you should see this screen:

.. image:: setup_assistant_panda_virtual_joints.png
   :width: 700px
   :align: center

Note: Virtual joints are particularly beneficial for robots with mobile bases,
such as mobile manipulators. They enable modeling the motion of the robot's base,
which is essential for motion planning and control. For instance, a virtual planar
joint can be used to connect the robot base frame to the odometry frame,

> effectively representing the robot's movement in the environment.

Reviewer: Consider adding an actual RST note here, like:

```
.. note::

   Virtual joints ...
```

Step 4: Add Planning Groups
---------------------------

Planning groups in MoveIt semantically describe different parts of the robot,
such as the arm or end effector, to facilitate motion planning.

A move group can be configured to correspond to a specific kinematic chain on the robot,
which is a set of links and joints that define a sequence of transformations
from the base of the robot to the end effector. For example, a move group might be defined to
represent the arm of a robot, which would consist of all the links and joints necessary to move the arm.

Move groups can also be represented by sets of links or joints corresponding on the robot.
For example, a move group might be defined to represent the gripper of a robot,

> which would consist of all the links or joints necessary to move together to realize a gripper opening motion.

Reviewer: ```suggestion
which would consist of all the links or joints necessary to move together to realize a gripper opening or closing motion.
```

* Click on the **Planning Groups** pane selector.

* Click on **Add Group** and you should see the following screen:

.. image:: planning_groups/setup_assistant_panda_planning_groups.png
   :width: 700px
   :align: center

Add the arm group

* We will first add Panda arm as a planning group

  * Enter **Group Name** as ``panda_arm``.

>   * Choose **kdl_kinematics_plugin/KDLKinematicsPlugin** as the
>     kinematics solver. Note: if you have a custom robot and would
>     like a powerful custom IK solver, see :doc:`IKFast Kinematics SolverIKFast </doc/examples/ikfast/ikfast_tutorial>` and :doc:`TRAC-IK Kinematics Solver </doc/examples/trac_ik/trac_ik_tutorial>`.

Reviewer: I'd reword this to be a little less tied to custom robots or "power" of an IK solver.

Also, `trac_ik` isn't yet available in ROS 2... but that will change soon if all goes well :)

```suggestion
  * Choose **kdl_kinematics_plugin/KDLKinematicsPlugin** as the kinematics solver, as this is the default for MoveIt.
    Alternatively, you can use other plugins such as :doc:`IKFast </doc/examples/ikfast/ikfast_tutorial>` or `pick_ik <https://github.com/PickNikRobotics/pick_ik>`_.
```

   :align: center

* We will also add a group for the end
  effector. Note that the end effector is not made of links attached in a serial chain.
  Therefore, the group's **Kinematic Solver** should be set to **None**.

  * Click on the **Add Group** button.

  * Enter **Group Name** as ``hand``.

  * Let **Kinematic Solver** stay at its default value, which is **None**.

  * Let **Kin. Search Resolution** and **Kin. Search Timeout** stay at their default values.

  * Click on the **Add Links** button.

  * Choose ``panda_hand``, ``panda_leftfinger``, and ``panda_rightfinger`` and add them
    to the list of **Selected Links** on the right-hand side.

  * Click **Save**.

.. image:: planning_groups/setup_assistant_panda_hand_group_links.png
   :width: 700px
   :align: center

After both arm and hand groups are added, the custom groups list should look as follows.

.. image:: planning_groups/setup_assistant_panda_planning_groups_done.png
   :width: 700px
   :align: center

Note: It is possible to build move groups composed of other move groups with the **Add Subgroup** option.
This can be beneficial in cases where multiple move groups need to be controlled together,

> such as when planning for simultaneous motions of multi-arm systems.

Reviewer: Same here, can use the `.. note::` admonition

Step 5: Add Robot Poses
-----------------------

> The Setup Assistant allows us to add predefined poses to the robot's configuration,

Reviewer: ```suggestion
The Setup Assistant allows you to add predefined poses to the robot's configuration,
```

> The Setup Assistant allows us to add predefined poses to the robot's configuration,

Reviewer: ```suggestion
The Setup Assistant allows you to add predefined poses to the robot's configuration,
```

then click the **Add Interfaces** button.

Step 9: ROS 2 Controllers
-------------------------

ROS 2 Control is a framework for real-time control of robots,
designed to manage and simplify the integration of new robot hardware.
For more details, please look at `ros2_control <https://control.ros.org/master/index.html>`_ documentation.

**ROS 2 Controllers** pane can be used to auto generate simulated controllers to actuate the robot joints.

.. image:: ros2_controllers/setup_assistant_ros2_controllers.png
   :width: 700px

Add the arm controllers

* Click on the **ROS 2 Controllers** pane selector.

* Click on **Add Controller** and you should see the following screen:

* We will first add Panda arm joint trajectory controller.

* Enter **Controller Name** as ``panda_arm_controller``.

* Choose **joint_trajectory_controller/JointTrajectoryController** as the controller type

.. image:: ros2_controllers/setup_assistant_panda_arm_ros2_controller_type.png
   :width: 700px

* Next, we need to choose the controller joints. Joints can be added individually or by move group.

* Now, click on **Add Planning Group Joints**.

> * Choose ``panda_arm`` group from the **Available Groups** tab and add it to the **Selected Groups**.

Reviewer: ```suggestion
* Choose the ``panda_arm`` group from the **Available Groups** tab and add it to the **Selected Groups**.
```

* Click **Save** to save the selected controller.

.. image:: ros2_controllers/setup_assistant_panda_arm_ros2_controller_group.png
   :width: 700px

Add the hand controllers

> * Follow the same steps for the arm, but choose **position_controllers/GripperActionController**

Reviewer: Instead of saying "follow the same steps for the arm", should we just list out the steps? Because notably what is missing from this section is what name to give this new controller (people can read it from the screenshot, I guess, but still...)

Author: Yes, this would be improve the clarity. I listed the steps. Thanks for the suggestion!

.. image:: ros2_controllers/setup_assistant_hand_ros2_controller_type.png
   :width: 700px

* Choose ``hand`` group from the **Available Groups** tab and add it to the **Selected Groups**.

* Click **Save** to save the selected controller.

.. image:: ros2_controllers/setup_assistant_hand_ros2_controller_group.png
   :width: 700px

After selecting the arm and hand controllers, the controllers list should be as follows.

.. image:: ros2_controllers/setup_assistant_ros2_controllers_done.png
   :width: 700px

Step 10: MoveIt Controllers
---------------------------

MoveIt requires trajectory controllers with a ``FollowJointTrajectoryAction`` interface for
executing planned trajectories. This interface sends the generated trajectory to the robot ROS 2 Controllers.

> **MoveIt Controllers** pane can be used to auto-generate the controllers to be used by the MoveIt controller manager.

Reviewer: ```suggestion
The **MoveIt Controllers** pane can be used to auto-generate the controllers to be used by the MoveIt controller manager.
```

* Choose the controller joints with the ``panda_arm`` planning group.

.. image:: moveit_controllers/setup_assistant_panda_arm_moveit_controller_type.png
   :width: 700px

Add the hand MoveIt controllers

* Follow the same steps for the arm, but choose **Gripper Command** Controller Type.

* Add the joints using the ``hand`` move group and save the controller.

.. image:: moveit_controllers/setup_assistant_hand_moveit_controller_type_gripper.png
   :width: 700px

After completing the previous steps, the MoveIt Controllers list for the arm and hand should appear as follows.

.. image:: moveit_controllers/setup_assistant_moveit_controllers_done_gripper.png
   :width: 700px

Step 11: Perception
-------------------

The Perception tab in the Setup Assistant is used to configure the settings
for 3D sensors used by the robot. These settings are saved in a YAML configuration file named **sensors_3d.yaml**.

In case of **sensors_3d.yaml** was not needed, choose **None** and proceed to the next step.

.. image:: perception/setup_assistant_panda_3d_perception.png
   :width: 700px

To generate **point_cloud** configuration parameters, see the following example:

> NOTE: this configuration is not valid for the Panda robot arm because it does not have a ``head_mounted_kinect`` camera.

Reviewer: Switch to `.. note::` admonition here too

.. image:: perception/setup_assistant_panda_3d_perception_point_cloud.png
   :width: 700px

For more details about those parameters please refer to the :doc:`Perception Pipeline tutorial </doc/examples/perception_pipeline/perception_pipeline_tutorial>`.

Step 12: Launch Files
---------------------

In the **Launch Files** pane, you can view the list of launch files that will be generated.
The default options are usually sufficient, but if you have specific requirements for your application,
you can make changes as necessary. Click on each of the files to view a summary of their functionality.

.. image:: setup_assistant_launch_files.png
   :width: 700px

Step 13: Add Author Information
--------------------------------

> Catkin requires author information for publishing purposes.

Reviewer: ```suggestion
Colcon requires author information for publishing purposes.
```

command the robot to move to the pre-defined ``ready`` pose and execute ``open`` and ``close`` motions on the hand.

What's Next
-----------

Get Started with MoveIt Motion Planning using RViz

* Learn how to use the generated configuration files to plan and visualize motion with MoveIt in RViz.
  Check out the :doc:`MoveIt Quickstart in Rviz tutorial </doc/tutorials/quickstart_in_rviz/quickstart_in_rviz_tutorial>` for a step-by-step guide.

Write Your First C++ MoveIt Application

* Write your first C++ application using MoveIt with :doc:`this tutorial </doc/tutorials/your_first_project/your_first_project>`,
  and familiarize yourself with the ``MoveGroupInterface`` and use it to plan, execute, and visualize motion plans for your robot from :doc:`this example </doc/examples/move_group_interface/move_group_interface_tutorial>`.

URDF vs SRDF: Understand the Differences

* See the :doc:`URDF and SRDF </doc/examples/urdf_srdf/urdf_srdf_tutorial>` page for more
  details on the components of the URDF and SRDF mentioned in this tutorial.

Setup Custom Inverse Kinematics Solvers

* Faster IK solvers than the default KDL solver are available.

>   For more information, refer to the :doc:`IKFast Kinematics SolverIKFast </doc/examples/ikfast/ikfast_tutorial>` and :doc:`TRAC-IK Kinematics Solver </doc/examples/trac_ik/trac_ik_tutorial>`.

Reviewer: As per my previous comment on IK plugins, try use more neutral wording. Specifically,

* KDL is actually pretty fast already, so switching away from KDL is usually to get better performance especially in the face of constraints. IKFast is the actual faster exception because it can generate analytical solutions, but other solvers that use optimization such as TRAC-IK, Bio-IK, and pick_ik are actually slower.
* Also, TRAC-IK is not yet available in ROS 2 so I would advise against remove it and point to `pick_ik` instead if you'd like.

Reviewer 3: trac_ik and bio_ik are not slower than KDL! See https://bitbucket.org/traclabs/trac_ik/src/master/

Reviewer: oh wow, that's surprising considering TRAC-IK is literally KDL + an optimization based solver in parallel. I guess the authors made some improvements to the Orocos KDL package...

Author: TRAC-IK is an enhancement of KDL + an optimization based solver. This enhancement is referred to as KDL-RR (Random Restarts) by the authors. Unlike the default KDL solver, KDL-RR uses a time-based run loop instead of a fixed number of iterations. Additionally, it utilizes random restarts (seeds) to overcome local minima when detected.

@sea-bass, I updated the tutorial to use neutral wording when alternative IK solvers are mentioned, and took a note this point may need reconsideration after more IK plugins are available.

MoveIt Setup Assistant
========================

.. image:: setup_assistant_launch.png
   :width: 700px
   :align: center

Overview
--------
The MoveIt Setup Assistant is a graphical user interface for configuring any robot for use with MoveIt. Its primary function is generating a Semantic Robot Description Format (SRDF) file for your robot, which specifies additional information required by MoveIt such as planning groups, end-effectors, and various kinematic parameters. Additionally, it generates other necessary configuration files for use with the MoveIt pipeline.
To use the MoveIt Setup Assistant, you will need to have a URDF file for your robot.

> Once you have a URDF file, you can open the MoveIt Setup Assistant and import your URDF. This tutorial will guide you through a series of steps to configure various aspects of your robot, such as defining its kinematic structure, specifying planning groups and end-effectors, and collision checking related setting. To learn more about the URDF and SRDF, you can refer to the `URDF_SRDF_Overview <https://moveit.picknik.ai/humble/doc/examples/urdf_srdf/urdf_srdf_tutorial.html#>`_ page.

Reviewer: Please consider doing one sentence per line. Makes code review a lot easier!

Author: Thank you for your comments. I have made the necessary changes to address the issue. Sorry if the previous version was hard to review.

Getting Started
---------------

MoveIt and ROS 2

* Follow the instructions for `installing MoveIt <https://moveit.picknik.ai/humble/doc/tutorials/getting_started/getting_started.html>`_
  first if you have not already done that.

* We use the `moveit_resources_panda_description <https://github.com/ros-planning/moveit_resources/tree/humble/panda_description/urdf>`_

>   package for Humble, which was already included in your workspace if you followed the previous step.

Author: Thank you for the suggestion! That step is directly above. I rephrased the sentence for clarity.

by disabling collision checking for pairs of links on the robot that are known to be safe.
This is achieved by determining which pairs of links are always in collision, never in collision,
in collision in the robot's default position, or adjacent to each other on the kinematic chain.

You can set the sampling density, which determines how many random robot positions are checked for self-collision.
Higher densities require more computation time, while lower densities may result in disabling pairs that should not be disabled.
By default, the generator checks 10,000 random positions for self-collision.

To speed up collision checking, the processing is done in parallel.
This reduces the overall processing time for generating the collision matrix.

**To generate the self-collision matrix**, first select the Self-Collisions pane on the left-hand side of the MoveIt Setup Assistant.
Here, you can adjust the self-collision sampling density to balance between computation time and accuracy of the results.

Once you have selected the desired sampling density, click on the Generate Collision Matrix button to initiate the computation.

> The Setup Assistant will take a few seconds to compute the self-collision matrix, which involves checking for pairs of links that can be safely disabled from collision checking.

Author: Yes, thank you for this suggestion! The step was updated accordingly.

.. image:: collision_matrix/setup_assistant_panda_collision_matrix.png
   :width: 700px
   :align: center

Once the computation is complete, the results will be presented in the main table. The table shows the pairs of links that have been identified as either safe or unsafe to disable from collision checking.

> Links that are safe to disable are marked with a checkmark. You can manually adjust the checkmarks as needed to enable or disable self-collision checking for specific link pairs

Reviewer: Missing full stop at the end of this sentence

.. image:: collision_matrix/setup_assistant_panda_collision_matrix_done.png
   :width: 700px
   :align: center

Step 3: Add Virtual Joints
--------------------------
Virtual joints are used primarily to attach the robot to the world.
For the Panda, we will define only one virtual joint attaching the *panda_link0*

> of the Panda to the *world* world frame. This virtual joint represents the motion of the base of the robot in a plane.

Author: I am truly thankful for yout comment, @MarqRazz!

My statement is wrong. It should be a fixed virtual joint, and it is optional because the Panda arm has a fixed base. I have also used single and multiple arms URDFs without virtual joints and the models worked as expected.

I did some searches and checked older documentations. My conclusion is as follows.

1. A fixed virtual joint provides fixes a robot base to the world. Here is a relevant [documentation](http://docs.ros.org/en/kinetic/api/moveit_tutorials/html/doc/urdf_srdf/urdf_srdf_tutorial.html#virtual-joints).

2. A planar virtual joint provides 3 DOF and attaches a mobile base to the world (this would be the one to use for mobile manipulators). Here is relevant [documentation](http://docs.ros.org/en/indigo/api/moveit_tutorials/html/doc/setup_assistant/setup_assistant_tutorial.html#step-3-add-virtual-joints) and a screenshot from [Mobile Manipulation workshop by PickNik team members](https://docs.google.com/presentation/d/1Qx66OICZrTRzJ73qj3tsJS0eOS2Cqw5e-JaNmpRGywQ/edit#slide=id.gf940579c23_3_6), page 13. 

![Planar joint for Stretch mobile manipulator](https://i.imgur.com/IFPiTsT.png)

3. A floating virtual joint provides 6 DOF and would be the one to use if we attach an arm to a quad-rotor for example. Here is a relevant [issue](https://github.com/ros-planning/moveit/issues/2111). 

I think the moveit_config in the resources package should use a fixed joint. I updated the tutorial based on this information. 

@sea-bass, can you please share with us how MoveIt uses the virtual joint and its advantages/disadvantages?

.. image:: setup_assistant_panda_virtual_joints.png
   :width: 700px
   :align: center

The virtual joint feature is especially useful when we have a robot that is attached to a mobile base,
such as a mobile manipulator. In such cases, the virtual joint allows us to model the motion

> of the robot's base, which is important for motion planning and control.

Author: Yes, you are right. I think the virtual joint will be between the mobile base and the odometry frame. I am not sure though.

> of the robot's base, which is important for motion planning and control.

Author: I will mark this as resolved to continue the discussion on the other relevant comment.

such as the arm or end effector, to facilitate motion planning.

A move group can be configured to correspond to a specific kinematic chain on the robot,
which is a set of links and joints that define a sequence of transformations from the base of the robot to the end effector.
For example, a move group might be defined to represent the arm of a robot, which would consist of all the links and joints necessary to move the arm.

Move groups can also be represented by sets of links or joints corresponding on the robot.
For example, a move group might be defined to represent the gripper of a robot, which would consist of all the links or joints necessary to move together to realize a gripper opening motion.

* Click on the *Planning Groups* pane selector.

* Click on *Add Group* and you should see the following screen:

.. image:: planning_groups/setup_assistant_panda_planning_groups.png
   :width: 700px
   :align: center

Add the arm group

* We will first add Panda arm as a planning group

  * Enter *Group Name* as **panda_arm**

  * Choose *kdl_kinematics_plugin/KDLKinematicsPlugin* as the
    kinematics solver. *Note: if you have a custom robot and would

>     like a powerful custom IK solver, see* `Kinematics/IKFast <https://moveit.picknik.ai/humble/doc/examples/ikfast/ikfast_tutorial.html>`_

Reviewer: Please continue using `:doc:` tags as these will automatically pick the right version. Do not hard-code a humble specific link.

  how poses are associated with particular groups. You can save
  individual poses for each group.

> * Select the *panda_arm* and define a **ready** pose for it with the following joint values {0.0, -0.785, 0.0, -2.356, 0.0, 1.571, 0.785}.

Reviewer: Use code font (double backticks) for the joint values

Author: Fixed it. Thanks for pointing this out!

* **IMPORTANT TIP**: Try to move all the joints around. If there is
  something wrong with the joint limits in your URDF, you should be able
  to see it immediately here.

.. image:: predefined_poses/setup_assistant_panda_predefined_arm_pose.png
   :width: 700px

Add *open* and *close* poses for the gripper

> * Follow the same steps for defning a pose for the arm, but select the *hand* group. 

Reviewer: small typo

```suggestion
* Follow the same steps for defining a pose for the arm, but select the *hand* group. 
```

* Add an *open* pose for the hand group with value 0.035

.. image:: predefined_poses/setup_assistant_panda_predefined_hand_open_pose.png
   :width: 700px

* Add a *close* pose for the hand with value 0.0

.. image:: predefined_poses/setup_assistant_panda_predefined_hand_close_pose.png
   :width: 700px

> Note: Only panda_finger_joint1 appears in the list as joint2 mimics its values.

Reviewer: Use code font for the joint names

> After the previous steps, the following set of robot poses should be defined for the panda_arm and hand groups.

Reviewer: Use code font for the group names

* Leave *Parent Group* blank.

.. image:: setup_assistant_panda_add_end_effector.png
   :width: 700px

Step 7: Add Passive Joints
--------------------------

The passive joints pane is meant to allow specification of any passive
joints that might exist in a robot. These are joints that are unactuated, 
meaning that they cannot be directly controlled. It's important to specify
passive joints so that the planners are aware of their existence and can avoid
planning for them. If the planners do not know about the passive joints, they 
might try to plan trajectories that involve moving the passive joints, which would
result in invalid plans. The Panda robot arm does not have any passive joints so we will skip this step.

Step 8: ros2_control URDF Modification
--------------------------------------

The ros2_control URDF Modification helps modify the robot URDF to work with 
`ros2_control <https://control.ros.org/master/index.html>`_.

This modification adds tags for command and state interfaces for each joint in the defined move groups. 
The command_interface tags define the types of commands that can be sent to control the joint.

> The state_interface tags define the types of state information that can be read from the joint.

Reviewer: code font for `command_interface` and `state_interface`

> By default the Setup Assistant assumes **position** command interface 

Reviewer: ```suggestion
By default the MoveIt Setup Assistant assumes **position** command interface 
```

> and **position** and **velocity** state interfaces, and we will proceed with this setting. 

Reviewer 2: FYI ros2_control does not support setting a [state_interface](https://github.com/ros-controls/ros2_controllers/blob/master/joint_trajectory_controller/src/joint_trajectory_controller_parameters.yaml#L28-L36) of `effort` right now. I did this and the resulting errors are extremely hard to understand and don't tell the user how to fix it! I will work on making a fix over there but wanted to point it out here.

Author: Understood. I appreciate your guidance on this issue!

.. image:: ros2_control/setup_assistant_ros2_control_tags.png
   :width: 700px

If necessary, select the desired command or state interfaces for your robot joints and 
then click the **Add Interface** button.

Step 9: ROS 2 Controllers
-------------------------

> The ROS 2 control is a framework for (real-time) control of robots using. 

Reviewer: ```suggestion
ROS 2 control is a framework for (real-time) control of robots using. 
```

also this sentence ends abruptly. Using what?

Author: Sorry, this was a mistake. I rewrote it.

ROS 2 Control is designed to manage and simplify the integration of new robot hardware. 

> For more details please look at `ros2_control <https://control.ros.org/master/index.html>`_ documentation.

Reviewer: ```suggestion
For more details please look at he `ros2_control <https://control.ros.org/master/index.html>`_ documentation.
```

ROS 2 Controllers tab can be used to auto generate simulated controllers to actuate the robot joints.

.. image:: ros2_controllers/setup_assistant_ros2_controllers.png
   :width: 700px

Add the arm controllers

* Click on the *ROS 2 Controllers* pane selector.

* Click on *Add Controller* and you should see the following screen:

* We will first add Panda arm joint trajectory controller

* Enter *Controller Name* as **panda_arm_controller**

* Choose **joint_trajectory_controller/JointTrajectoryController** as the controller type

.. image:: ros2_controllers/setup_assistant_panda_arm_ros2_controller_type.png
   :width: 700px

> * Next, we need to choose this controller joints. Joints can be added individually or by move group.

Reviewer: Either "the controller joints" or "this controller's joints"

* Now, click on Add Planning Group Joints.

> * Choose panda_arm planning group to add all the joints in that group to the arm controller.

Reviewer: Code font for `panda_arm`

Reviewer 2: Choose the `panda_arm` group from the `Available Groups` tab and add it to the `Selected Groups`.

Author: Thanks you for your comments @sea-bass and @MarqRazz.

For consistency with using bold font with GUI menus/items, I made it as follows.

Choose the `panda_arm` from the **Available Groups** and add it to the **Selected Groups**.

* Click *Save* to save the selected controller.
  
.. image:: ros2_controllers/setup_assistant_panda_arm_ros2_controller_group.png
   :width: 700px

Add the hand controllers

> * Follow the same steps for the arm, but choose **position_controllers/GripperActionCOntroller** 

Reviewer 2: Typo, should be: **position_controllers/GripperActionController**

Author: Thank you! This typo is fixed.

.. image:: ros2_controllers/setup_assistant_hand_ros2_controller_type.png
   :width: 700px

> * Then, add the joints using the hand move group and save the controller.

Reviewer 2: * Choose the `hand` group from the `Available Groups` tab and add it to the `Selected Groups`.
* Click *Save* to save the selected controller.

Author: Thank you for your feedback! The steps have been updated to ensure consistency.

.. image:: ros2_controllers/setup_assistant_hand_ros2_controller_group.png
   :width: 700px

> After selecting the arm and hand controllers, the controllers list should be as follows

Reviewer: Add `.` or `:` at end of sentence

.. image:: ros2_controllers/setup_assistant_ros2_controllers_done.png
   :width: 700px

Step 10: MoveIt Controllers
---------------------------

> MoveIt requires trajectory controllers with a FollowJointTrajectoryAction interface for 

Reviewer: code font for `FollowJointTrajectoryAction`

executing planned trajectories. This interface sends the generated trajectory to the robot ROS 2 Controllers.

> MoveIt Controllers pane can be used to auto-generate the controllers to be used by MoveIt controller manager. 

Reviewer 2: controllers to be used by `the` MoveIt controller manager.

> Ensure that the controller names match those configured in the previous  ROS 2 controller step. 

Reviewer: extra double space

```suggestion
Ensure that the controller names match those configured in the previous ROS 2 controller step. 
```

* Enter Controller Name as panda_arm_controller

* Choose **FollowJointTrajectory** Controller Type. 

* Choose the controller joints with the panda_arm planning group.

.. image:: moveit_controllers/setup_assistant_panda_arm_moveit_controller_type.png
   :width: 700px

Add the hand MoveIt controllers

* Follow the same steps for the arm, but choose **Gripper Command** Controller Type. 
  
* Add the joints using the hand move group and save the controller.

.. image:: moveit_controllers/setup_assistant_hand_moveit_controller_type_gripper.png
   :width: 700px
  
After completing the previous steps, the MoveIt Controllers list for the arm and hand should appear as follows.

.. image:: moveit_controllers/setup_assistant_moveit_controllers_done_gripper.png
   :width: 700px

Step 11: Perception
-------------------

The Perception tab in the Setup Assistant is used to configure the settings 
for 3D sensors used by the robot. These settings are saved in a YAML configuration file named **sensors_3d.yaml**.

In case of **sensors_3d.yaml** was not needed, choose **None** and proceed to the next step.

.. image:: perception/setup_assistant_panda_3d_perception.png
   :width: 700px

> To generate **point_cloud** configuration parameters:

Reviewer 2: To generate **point_cloud** configuration parameters see the following example:
NOTE: this configuration is not valid for the Panda robot arm because it does not have a `head_mounted_kinect` camera.

Author: Thank you for highlighting this point! I updated the perception step accordingly.

.. image:: perception/setup_assistant_panda_3d_perception_point_cloud.png
   :width: 700px

For more details about those parameters please refer to the `perception pipeline tutorial <https://moveit.picknik.ai/humble/doc/examples/perception_pipeline/perception_pipeline_tutorial.html>`_.

> 

Reviewer: Why all these spaces?

Author: Removed unneeded spaces.

Step 12: Launch Files
---------------------

In the *Launch Files* pane, you can view the list of launch files that will be generated. 
The default options are usually sufficient, but if you have specific requirements for your application, 
you can make changes as necessary. Click on each of the files to view a summary of their functionality.

.. image:: setup_assistant_launch_files.png
   :width: 700px

Step 13: Add Author Information
--------------------------------

> Catkin requires author information for publishing purposes

Reviewer 2: ROS packages require author information for publishing purposes.

   :width: 700px

Congratulations! You are now done generating the configuration files you need for MoveIt.

Build the panda_moveit_config package and run the demo
------------------------------------------------------
To build only the generated panda_moveit_config package and run the demo, follow these steps. ::

   cd ~/ws_moveit2
   colcon build --packages-select panda_moveit_config
   source install/setup.bash

Start the MoveIt demo to interactively plan and execute motions for the robot in RViz. :: 

   ros2 launch panda_moveit_config demo.launch.py

Check out this `brief YouTube video <https://youtu.be/CXeF5KRa55Y>`_ for an example of how to command the robot to move to the pre-defined ready pose and execute open and close motions on the hand.

What's Next
-----------

Get Started with MoveIt Motion Planning using RViz

* Learn how to use the generated configuration files to plan and visualize motion with MoveIt in RViz. 
  Check out the `MoveIt Quickstart in Rviz tutorial <https://moveit.picknik.ai/humble/doc/tutorials/quickstart_in_rviz/quickstart_in_rviz_tutorial.html>`_ for a step-by-step guide.

Write Your First C++ MoveIt Application

> * Write your first C++ application using MoveIt with this `tutorial <https://moveit.picknik.ai/humble/doc/tutorials/your_first_project/your_first_project.html>`_, and familiarize yourself with the MoveGroupInterface and use it to plan, execute, and visualize motion plans for your robot from this `example <https://moveit.picknik.ai/humble/doc/examples/move_group_interface/move_group_interface_tutorial.html>`_. 

Reviewer: Make sure the "this" is also part of the hyperlinks

URDF vs SRDF: Understand the Differences

* See the `URDF and SRDF <https://moveit.picknik.ai/humble/doc/examples/urdf_srdf/urdf_srdf_tutorial.html>`_ page for more
  details on the components of the URDF and SRDF mentioned in this tutorial.

Setup IKFast Inverse Kinematics Solver

* A faster IK solver than the default KDL solver, but takes some

>   additional steps to setup `Kinematics/IKFast <https://moveit.picknik.ai/humble/doc/examples/ikfast/ikfast_tutorial.html>`_

Reviewer: This link doesn't quite fit in the full sentence. Consider rewording or adding a : in between

Author: Thank you for the feedback. I have rephrased this sentence, but I would appreciate your opinion on whether it is suitable to keep. I took this point from the old documentation.

Author: I updated this sentence again and added a link to TRAC-IK.
